File: scripts/sender.py
#!/usr/bin/env python3
"""Standalone sender that auto-connects via relay (default) or Corelink."""

from __future__ import annotations

import argparse
import asyncio
import sys

import requests


def fetch_credentials(host: str, deploy_id: str) -> dict:
    url = f"{host}/deployments/{deploy_id}/credentials"
    resp = requests.get(url, params={"role": "sender"}, timeout=10)
    if resp.status_code != 200:
        print(f"Error fetching credentials: {resp.status_code} {resp.text}", file=sys.stderr)
        sys.exit(1)
    return resp.json()


async def run_relay(host: str, deploy_id: str) -> None:
    """Send messages through the backend HTTP relay (no Corelink needed)."""
    url = f"{host}/deployments/{deploy_id}/messages"

    print(f"Sender connected via relay (deployment: {deploy_id})")
    print("Type messages to send (Ctrl+C to quit):\n")

    try:
        while True:
            line = await asyncio.get_event_loop().run_in_executor(
                None, lambda: input("> ")
            )
            if not line:
                continue
            resp = requests.post(url, json={"data": line}, timeout=10)
            if resp.status_code == 200:
                info = resp.json()
                print(f"  sent ({info.get('listeners', 0)} listeners)")
            else:
                print(f"  error: {resp.status_code} {resp.text}", file=sys.stderr)
    except (KeyboardInterrupt, EOFError):
        print("\nDone.")


async def run_corelink(host: str, deploy_id: str) -> None:
    """Send messages through a live Corelink server."""
    import corelink  # noqa: delayed import — only needed in corelink mode

    data = fetch_credentials(host, deploy_id)
    creds = data["credentials"]

    if not creds:
        print("No sender credentials found in this deployment.", file=sys.stderr)
        sys.exit(1)

    stream_type = next(iter(creds))
    cred = creds[stream_type]
    workspace = cred["workspace"]
    protocol = cred.get("protocol", "pubsub")

    print(f"Connecting to Corelink as sender...")
    print(f"  Workspace : {workspace}")
    print(f"  Data type : {stream_type}\n")

    await corelink.connect("Corelink", "Corelink2023")
    sender = await corelink.create_sender(
        workspace=workspace, protocol=protocol, data_type=stream_type,
    )

    print("Connected! Type messages to send (Ctrl+C to quit):\n")

    try:
        while True:
            line = await asyncio.get_event_loop().run_in_executor(
                None, lambda: input("> ")
            )
            if not line:
                continue
            await sender.send(line.encode("utf-8"))
            print(f"  sent: {line}")
    except (KeyboardInterrupt, EOFError):
        print("\nDisconnecting...")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Auto-connect sender for a deployment")
    parser.add_argument("deploy_id", help="Deployment ID from the orchestrator")
    parser.add_argument(
        "--host", default="http://localhost:8000",
        help="Backend URL (default: http://localhost:8000)",
    )
    parser.add_argument(
        "--mode", choices=["relay", "corelink"], default="relay",
        help="Message transport (default: relay — no Corelink server needed)",
    )
    args = parser.parse_args()

    if args.mode == "corelink":
        asyncio.run(run_corelink(args.host, args.deploy_id))
    else:
        asyncio.run(run_relay(args.host, args.deploy_id))

File: scripts/sender.js
#!/usr/bin/env node
/**
 * Standalone sender that auto-connects via NATS (default) or relay.
 * Usage: node sender.js <deploy_id> [--mode nats|relay] [--host URL] [--nats-host HOST]
 */

const readline = require('readline')

function parseArgs(argv) {
  const args = { mode: 'nats', host: 'http://localhost:8000', natsHost: null }
  const positional = []
  for (let i = 2; i < argv.length; i += 1) {
    const a = argv[i]
    if (a === '--mode') args.mode = argv[++i]
    else if (a === '--host') args.host = argv[++i]
    else if (a === '--nats-host') args.natsHost = argv[++i]
    else if (a.startsWith('--')) { console.error(`unknown flag: ${a}`); process.exit(2) }
    else positional.push(a)
  }
  if (positional.length !== 1) {
    console.error('usage: sender.js <deploy_id> [--mode nats|relay] [--host URL] [--nats-host HOST]')
    process.exit(2)
  }
  args.deployId = positional[0]
  return args
}

async function fetchCredentials(host, deployId, role) {
  const url = `${host}/deployments/${deployId}/credentials?role=${role}`
  const resp = await fetch(url)
  if (!resp.ok) {
    console.error(`Error fetching credentials: ${resp.status} ${await resp.text()}`)
    process.exit(1)
  }
  return resp.json()
}

async function main() {
  const args = parseArgs(process.argv)
  const transport = args.mode === 'nats'
    ? require('./lib/nats-transport')
    : require('./lib/relay-transport')

  console.log(`Mode: ${args.mode}`)
  const cred = await fetchCredentials(args.host, args.deployId, 'sender')

  // Allow CLI override of the nats host (useful when the plugin container
  // and the host process need different hostnames to reach the same broker).
  const natsBlock = cred.nats && args.natsHost
    ? { ...cred.nats, host: args.natsHost, url: `nats://${args.natsHost}:${cred.nats.port}` }
    : cred.nats

  let handle
  try {
    handle = await transport.connect({
      host: args.host,
      deployId: args.deployId,
      role: 'sender',
      credentials: cred.credentials,
      natsBlock,
    })
  } catch (e) {
    console.error(`Connect failed (${args.mode}): ${e.message}`)
    if (args.mode === 'nats') console.error('Try --mode relay as a fallback.')
    process.exit(1)
  }

  console.log(`Sender connected (deployment: ${args.deployId})`)
  console.log('Type messages to send (Ctrl+C to quit):\n')

  const rl = readline.createInterface({ input: process.stdin, output: process.stdout })
  rl.setPrompt('> ')
  rl.prompt()
  rl.on('line', async (line) => {
    if (line) {
      try {
        await transport.send(handle, line)
        console.log(`  sent: ${line}`)
      } catch (e) {
        console.error(`  send error: ${e.message}`)
      }
    }
    rl.prompt()
  })
  rl.on('close', async () => {
    await transport.close(handle).catch(() => {})
    console.log('\nDone.')
    process.exit(0)
  })
}

main().catch((e) => { console.error(e); process.exit(1) })
